File: yabause/src/android/app/src/main/java/org/uoyabause/android/GameDirectoriesDialogFragment.java
package org.uoyabause.android;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.preference.PreferenceDialogFragmentCompat;

import org.devmiyax.yabasanshiro.R;
import org.uoyabause.android.tv.GameSelectFragment;

import java.io.File;
import java.util.ArrayList;

class DirectoryListAdapter extends BaseAdapter implements ListAdapter {

    private Context context;
    private ArrayList<String> _directoryList = new ArrayList<String>();

    ArrayList<String> getList(){ return _directoryList; }

    public DirectoryListAdapter(Context context) {
        this.context = context;
    }

    public void setDirectorytList(ArrayList<String> directoryList) {
        this._directoryList = directoryList;
    }

    @Override
    public int getCount() {
        return _directoryList.size();
    }

    @Override
    public Object getItem(int position) {
        return _directoryList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position; //_directoryList.get(position).getId();
    }

    public void addDirectory( String path ){
        _directoryList.add(path);
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view = convertView;
        if (view == null) {
            LayoutInflater inflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater.inflate(R.layout.game_directories_listitem, null);
        }
        ((TextView)view.findViewById(R.id.game_directory_item)).setText(_directoryList.get(position));

        Button deleteBtn = (Button)view.findViewById(R.id.button_delete);
        deleteBtn.setTag(position);
        deleteBtn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {
                Integer position = (Integer)v.getTag();
                if( position != null ){
                    _directoryList.remove(position.intValue()); //or some other task
                    notifyDataSetChanged();
                    GameSelectFragment.refresh_level = 3;
                }
            }
        });

        return view;
    }
}

public class GameDirectoriesDialogFragment extends PreferenceDialogFragmentCompat implements View.OnClickListener , FileDialog.DirectorySelectedListener {

    DirectoryListAdapter adapter;
    ListView listView;

    public static GameDirectoriesDialogFragment newInstance(String key) {
        GameDirectoriesDialogFragment fragment = new GameDirectoriesDialogFragment();
        Bundle b = new Bundle(1);
        b.putString(PreferenceDialogFragmentCompat.ARG_KEY, key);
        fragment.setArguments(b);
        return fragment;
    }

    @Override
    public View onCreateDialogView(Context context){
        View v =  super.onCreateDialogView(context);
        return v;
    }

    @Override
    protected void onPrepareDialogBuilder( AlertDialog.Builder builder) {
        super.onPrepareDialogBuilder(builder);
        LayoutInflater inflater = (LayoutInflater)getActivity().getLayoutInflater();
        View view = inflater.inflate(R.layout.game_directories, null, false);
        onBindDialogView(view);
        builder.setView(view);
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        if(positiveResult){
            String resultstring = "";
            ArrayList<String> list = adapter.getList();
            for( int i=0; i< list.size(); i++ ){
                resultstring += list.get(i);
                resultstring += ";";
            }
            GameDirectoriesDialogPreference is = (GameDirectoriesDialogPreference)getPreference();
            is.save(resultstring);
        }
    }

    @Override
    protected void onBindDialogView(View view) {
        super.onBindDialogView(view);
        ArrayList<String> list;
        list = new ArrayList<String>();

        GameDirectoriesDialogPreference is = (GameDirectoriesDialogPreference)getPreference();
        String data = is.getData();
        if (data.equals("err")) {
            YabauseStorage yb = YabauseStorage.getStorage();
            list.add(yb.getGamePath());
        } else {
            String[] paths = data.split(";", 0);
            for (int i = 0; i < paths.length; i++) {
                list.add(paths[i]);
            }
        }
        listView = (ListView) view.findViewById(R.id.listView);
        adapter = new DirectoryListAdapter(requireActivity());
        adapter.setDirectorytList(list);
        listView.setAdapter(adapter);
        View button = (Button) view.findViewById(R.id.button_add);
        button.setOnClickListener(this);
    }

    public void onClick(View v) {
        FileDialog fd = new FileDialog(requireActivity(),"");
        fd.setSelectDirectoryOption(true);
        fd.addDirectoryListener(this);
        fd.showDialog();
    }

    @Override
    public void directorySelected(File directory){
        adapter.addDirectory(directory.getAbsolutePath());
        adapter.notifyDataSetChanged();
        GameSelectFragment.refresh_level = 3;
    }
}

File: yabause/src/android/app/src/main/java/org/uoyabause/android/GameDirectoriesDialogFragment.kt
package org.uoyabause.android

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Button
import android.widget.ListAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.documentfile.provider.DocumentFile
import androidx.preference.PreferenceDialogFragmentCompat
import java.io.File
import org.devmiyax.yabasanshiro.R
import org.uoyabause.android.FileDialog.DirectorySelectedListener
import org.uoyabause.android.YabauseStorage.Companion.storage

class DirectoryListAdapter(private val context: Context) : BaseAdapter(), ListAdapter {
  var list = ArrayList<String>()
    private set

  fun setDirectorytList(directoryList: ArrayList<String>) {
    list = directoryList
  }

  override fun getCount(): Int {
    return list.size
  }

  override fun getItem(position: Int): Any {
    return list[position]
  }

  override fun getItemId(position: Int): Long {
    return position.toLong() // _directoryList.get(position).getId();
  }

  fun addDirectory(path: String) {
    list.add(path)
  }

  override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
    var view = convertView
    if (view == null) {
      val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
      view = inflater.inflate(R.layout.game_directories_listitem, null)
    }

    (view!!.findViewById<View>(R.id.game_directory_item) as TextView).text = list[position]
    val deleteBtn = view.findViewById<View>(R.id.button_delete) as Button
    deleteBtn.tag = position
    deleteBtn.setOnClickListener { v ->
      val clickPosition = v.tag as Int
      list.removeAt(clickPosition)
      notifyDataSetChanged()
    }
    return view
  }
}

class GameDirectoriesDialogFragment : PreferenceDialogFragmentCompat(), View.OnClickListener,
  DirectorySelectedListener {
  var adapter: DirectoryListAdapter? = null
  var listView: ListView? = null

  override fun onPrepareDialogBuilder(builder: AlertDialog.Builder) {
    super.onPrepareDialogBuilder(builder)
    val view = layoutInflater.inflate(R.layout.game_directories, null, false)

    val list: ArrayList<String>
    list = ArrayList()
    val `is` = preference as GameDirectoriesDialogPreference
    val data = `is`.data
    if (data == "err") {
      val yb = storage
      list.add(yb.gamePath)
    } else {
      val paths = data.split(";").dropLastWhile { it.isEmpty() }
        .toTypedArray()
      for (i in paths.indices) {
        list.add(paths[i])
      }
    }
    listView = view.findViewById<View>(R.id.listView) as ListView
    adapter = DirectoryListAdapter(requireActivity())
    adapter!!.setDirectorytList(list)
    listView!!.adapter = adapter
    val button: View = view.findViewById<View>(R.id.button_add) as Button
    button.setOnClickListener(this)

    builder.setView(view)
  }

  override fun onDialogClosed(positiveResult: Boolean) {
    if (positiveResult) {
      var resultstring = ""
      val list = adapter!!.list
      for (i in list.indices) {
        resultstring += list[i]
        resultstring += ";"
      }
      val `is` = preference as GameDirectoriesDialogPreference
      `is`.save(resultstring)
    }
  }

  override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
    super.onActivityResult(requestCode, resultCode, data)
    if (requestCode == 111 && resultCode == Activity.RESULT_OK) {

      val contentResolver = YabauseApplication.appContext.contentResolver
      val takeFlags: Int = Intent.FLAG_GRANT_READ_URI_PERMISSION
      contentResolver.takePersistableUriPermission(data!!.data!!, takeFlags)

      var cursor: Cursor? = null
      try {
        val pickedDir = DocumentFile.fromTreeUri(YabauseApplication.appContext, data.data!!)
        for (file in pickedDir!!.listFiles()) {
          Log.d("Yabause", "Found file " + file.name + " with size " + file.length())
        }
        adapter!!.addDirectory(data.data!!.toString())
        adapter!!.notifyDataSetChanged()
      } catch (e: Exception) {
        e.localizedMessage?.let { Log.e("Yabause", it) }
      } finally {
        cursor?.close()
      }
    }
  }

  override fun onClick(v: View) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {

      val rtn = YabauseApplication.checkDonated(requireActivity(), "")
      if (rtn == 0) {
        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE).apply {
          flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
        }
        startActivityForResult(intent, 111)
      }
    } else {
      val fd = FileDialog(requireActivity(), "")
      fd.setSelectDirectoryOption(true)
      fd.addDirectoryListener(this)
      fd.showDialog()
    }
  }

  override fun directorySelected(directory: File?) {
    if (directory != null) {
      adapter!!.addDirectory(directory.absolutePath)
      adapter!!.notifyDataSetChanged()
    }
  }

  companion object {
    fun newInstance(key: String?): GameDirectoriesDialogFragment {
      val fragment = GameDirectoriesDialogFragment()
      val b = Bundle(1)
      b.putString(ARG_KEY, key)
      fragment.arguments = b
      return fragment
    }
  }
}
